### DISCUSSION TITLE
InMemoryCache and shouldReadFromCache (apollo-server-plugin-response-cache)

### DISCUSSION THREAD

**Discussion by DmitriiKholodov** [apollographql/apollo-server#5271] [Q&A] [answered]
Hello! I have a question about InMemoryCache.

I am using the `shouldReadFromCache()` method from `apollo-server-plugin-response-cache` to check when I shouldn't use the cached data.

`shouldReadFromCache?(requestContext: GraphQLRequestContext<TContext>): ValueOrPromise<boolean>`

So when this method returns false - we make a new request with the same arguments. Is it means that the previous cached request will be deleted from the cache?

In continuation of my question - Is it possible to return the key of cached response and get access to methods of `InMemoryCache` class like `get()`, `set()`, `delete()`?

Best regards!

**Comment by glasser** [comment, marked as answer]
This means literally what it says: should the operation read from the cache or not? If it returns false, it never talks to the cache at all (no deletion or anything). You could use this to implement an HTTP request header that means "read a fresh value always", for example.

`GraphQLRequestContext` has a `cache` on it, but you don't have access to the cache key in `shouldReadFromCache`. (Part of the idea is to not take the performance hit of calculating the key in this case.)

It sounds like you want to add extra logic around talking to the cache. One option is that you could implement the small `KeyValueCache` interface yourself, wrapping and delegating to a real cache. Then if you want your `get` to also delete from the cache in some cases, you can have it do that.

**Reply by mmerezhko-hv** [reply]
Hello everyone! I have another question about the default in-memory cache of 30 MB (`cache: 'bounded'`).
- What happens if one request exceeds the cache limit?
- Does it reset between requests? Or does it not affect at all?

Thank you in advance!

**Reply by trevor-scheer** [reply]
The default cache uses the `lru-cache` node package under the hood. According to their docs, a single entry which exceeds the total cache size won't be cached (nothing will happen).
https://github.com/isaacs/node-lru-cache#maxsize-read-only 

Side note: if you're using Apollo Server v4, `cache: 'bounded'` is the default behavior and no longer needs to be specified.

**Reply by mmerezhko-hv** [reply]
Thank you!
